File: ios/RNPaypal/RNPaypal.m

#import "RNPaypal.h"
#import "React/RCTBridge.h"
#import "PayPalMobile.h"
#import "React/RCTConvert.h"

@interface RNPaypal () <PayPalPaymentDelegate, RCTBridgeModule>

@property(nonatomic, strong, readwrite) PayPalPayment *payment;
@property(nonatomic, strong, readwrite) PayPalConfiguration *config;
@property (copy) RCTResponseSenderBlock flowCompletedCallback;
@property (copy) RCTPromiseResolveBlock resolve;
@property (copy) RCTPromiseRejectBlock reject;
@property (readwrite, strong, nonatomic) NSArray *requiredParams;

@end


@implementation RNPaypal


- (dispatch_queue_t)methodQueue
{
    _requiredParams = [NSArray arrayWithObjects: @"clientId", @"environment", @"intent", @"price", @"currency", @"description", @"acceptCreditCards", nil];
    return dispatch_get_main_queue();
}
RCT_EXPORT_MODULE()


RCT_REMAP_METHOD(paymentRequest,
                 params:(NSDictionary *)params
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject){
    
    self.resolve = resolve;
    self.reject = reject;
    
    NSSet *requiredParamsSet = [NSSet setWithArray:_requiredParams];
    NSSet *paramsSet = [NSSet setWithArray:params.allKeys];
    
    if (![requiredParamsSet isEqualToSet:paramsSet]) {
        NSError *error = [NSError errorWithDomain:@"com.RNPaypal" code:-10 userInfo:@{@"Error reason": @"Invalid Input"}];
        self.reject(@"INVALID_PARAMS", @"Params are missing", error);
        return;
    }
    //
    //    if (![params.allKeys isEqualToArray:_requiredParams]) {
    //        NSError *error = [NSError errorWithDomain:@"com.RNPaypal" code:-10 userInfo:@{@"Error reason": @"Invalid Input"}];
    //        self.reject(@"INVALID_PARAMS", @"Params are missing", error);
    //    }
    //
    
    
    [self configureEnvironment:params];
    [self configurePaymentSettings:params];
    
    if(self.payment.processable) {
        PayPalPaymentViewController *paymentViewController = [[PayPalPaymentViewController alloc] initWithPayment:self.payment
                                                                                                    configuration:self.config
                                                                                                         delegate:self];
        UIViewController *visibleVC = [[[UIApplication sharedApplication] keyWindow] rootViewController];
        do {
            if ([visibleVC isKindOfClass:[UINavigationController class]]) {
                visibleVC = [(UINavigationController *)visibleVC visibleViewController];
            } else if (visibleVC.presentedViewController) {
                visibleVC = visibleVC.presentedViewController;
            }
        } while (visibleVC.presentedViewController);
        dispatch_async(dispatch_get_main_queue(), ^{
            [visibleVC presentViewController:paymentViewController animated:YES completion:nil];
        });
    }
    
}


-(void)configureEnvironment:(NSDictionary *) params {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString *envString = params[@"environment"];
        NSString *clientId = params[@"clientId"];
        
        [PayPalMobile initializeWithClientIdsForEnvironments:@{envString : clientId}];
        [PayPalMobile preconnectWithEnvironment:envString];
    });
}

-(void)configurePaymentSettings:(NSDictionary *) params {
    self.payment = [[PayPalPayment alloc] init];
    self.config = [[PayPalConfiguration alloc] init];

    _payment.amount = params[@"price"];
    _payment.intent = [params[@"intent"] intValue];
    _payment.currencyCode = params[@"currency"];
    _payment.shortDescription = params[@"description"];

    _config.acceptCreditCards = [params[@"acceptCreditCards"] boolValue];
}

- (NSDictionary *)constantsToExport
{
    return @{
             @"INTENT": @{
                     @"SALE":@0,
                     @"AUTHORIZE":@1,
                     @"ORDER":@2
                     
                     },
             @"ENVIRONMENT": @{
                     @"SANDBOX": PayPalEnvironmentSandbox,
                     @"PRODUCTION":PayPalEnvironmentProduction,
                     @"NO_NETWORK":PayPalEnvironmentNoNetwork
                     }
             };
}



- (void)payPalPaymentDidCancel:(nonnull PayPalPaymentViewController *)paymentViewController {
    [paymentViewController.presentingViewController dismissViewControllerAnimated:YES completion:^{
        if (self.reject) {
            NSError *error = [NSError errorWithDomain:@"com.RNPaypal" code:-40 userInfo:@{@"Error reason": @"Invalid Input"}];
            self.reject(@"USER_CANCELLED", @"User cancelled", error);
        }
    }];
}

- (void)payPalPaymentViewController:(nonnull PayPalPaymentViewController *)paymentViewController didCompletePayment:(nonnull PayPalPayment *)completedPayment {
    
    [paymentViewController.presentingViewController dismissViewControllerAnimated:YES completion:^{
        if(self.resolve) {
            self.resolve(completedPayment.confirmation);
        }
    }];
    
    
}

@end

File: ios/RNPaypal.swift
//
//  RNPaypal.swift
//  OneBoatApp
//
//  Created by Edward R on 1/11/18.
//  Copyright © 2018 Facebook. All rights reserved.
//

import UIKit


//RNPaypal.switf


@objc(RNPaypal)
class RNPaypal: UIViewController, PayPalPaymentDelegate, PayPalProfileSharingDelegate {
  func userDidCancel(_ profileSharingViewController: PayPalProfileSharingViewController) {
    
  }
  
  func payPalProfileSharingViewController(_ profileSharingViewController: PayPalProfileSharingViewController, userDidLogInWithAuthorization profileSharingAuthorization: [AnyHashable : Any]) {
    
  }
  
  
  var payPalConfig = PayPalConfiguration()
  var payPalPayment = PayPalPayment();
  var successCallback: RCTResponseSenderBlock? = nil;
  var errorCallback: RCTResponseSenderBlock? = nil;
  
 
//  var environment:String = PayPalEnvironmentProduction {
//    willSet(newEnvironment) {
//      if (newEnvironment != environment) {
//        PayPalMobile.preconnect(withEnvironment: newEnvironment)
//      }
//    }
//  }
  
  
 
  
  func payPalPaymentDidCancel(_ paymentViewController: PayPalPaymentViewController) {
  
    print("Payment canceled")
    paymentViewController.dismiss(animated: true, completion: nil)
    self.errorCallback!([["error_code": 101, "error_message" : "Payment Cancelled"], NSNull()])
  }
  
  func payPalPaymentViewController(_ paymentViewController: PayPalPaymentViewController, didComplete completedPayment: PayPalPayment) {
    
    paymentViewController.dismiss(animated: true, completion: { () -> Void in
       self.successCallback!([completedPayment.confirmation])
    })
    
  }
  
  fileprivate func configureEnvironment(_ payPalParameters: [String : Any]) {
    if let clientId = payPalParameters["clientId"] as? NSString {
      
      if let environment = payPalParameters["environment"] as? NSString {
        DispatchQueue.main.async {
          PayPalMobile.initializeWithClientIds(forEnvironments: [environment : clientId ])
          PayPalMobile.preconnect(withEnvironment: environment as String)
        }
      }
      
    }
  }
  
  fileprivate func configurePaymentSettings(_ payPalParameters: [String : Any]) {
    if let intent = payPalParameters["intent"] as? NSNumber {
      payPalPayment.intent = PayPalPaymentIntent(rawValue: Int(intent))!;
    }
    
    if let acceptCreditCards = payPalParameters["acceptCreditCards"] as? Bool {
      payPalConfig.acceptCreditCards = acceptCreditCards;
    }
    
    if let price = payPalParameters["price"] as? NSNumber {
      payPalPayment.amount = NSDecimalNumber(decimal: price.decimalValue)
    }
    
    if let currency = payPalParameters["currency"] as? String {
      payPalPayment.currencyCode = currency
    }
    
    if let description = payPalParameters["description"] as? String {
      payPalPayment.shortDescription = description
    }
  }
  
  @objc(paymentRequest:payPalParameters:successCallback:errorCallback:)
  func paymentRequest(name: String, payPalParameters: [String: Any], successCallback: @escaping RCTResponseSenderBlock, errorCallback: @escaping RCTResponseSenderBlock) -> Void {
    
    self.successCallback = successCallback;
    self.errorCallback = errorCallback;

    configureEnvironment(payPalParameters)
    
    configurePaymentSettings(payPalParameters)


    if (payPalPayment.processable) {
       let paymentViewController = PayPalPaymentViewController(payment: payPalPayment, configuration: payPalConfig, delegate: self)
       DispatchQueue.main.async {
        let rootVC: UIViewController? = UIApplication.shared.keyWindow?.rootViewController
        rootVC?.present(paymentViewController!, animated: true, completion: nil)
      }
     
    }
    else {
       self.errorCallback!([["error_code": 102, "error_message" : "Payment not processable"], NSNull()])
    }
    
  }

  
  func constantsToExport() -> [AnyHashable: Any]! {
    return [
      "ENVIRONMENT": [
        "SANDBOX": PayPalEnvironmentSandbox,
        "PRODUCTION": PayPalEnvironmentProduction,
        "NO_NETWORK": PayPalEnvironmentNoNetwork,
      ],
      "INTENT": [
        "SALE": PayPalPaymentIntent.sale,
        "AUTHORIZE": PayPalPaymentIntent.authorize,
        "ORDER": PayPalPaymentIntent.order
      ]
      
    ]
  }
  
  
}

